plugins/TauMVAProducer.cc

```diff
@@ -0,0 +1,371 @@
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/Handle.h"
+
+#include "DataFormats/Common/interface/Ptr.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/Common/interface/PtrVector.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+
+#include "TLorentzVector.h"
+
+class TauMVAProducer : public edm::stream::EDProducer<> 
+{
+public:
+    explicit TauMVAProducer(const edm::ParameterSet&);
+    ~TauMVAProducer();
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+    bool  isA(int particleID, int pdgId, bool checkCharge = false);
+    int   getContainingJetIndex(const pat::PackedCandidate* pfc) const;
+    int   getNearPhotonIndex(const pat::PackedCandidate* pfc);
+    float getDRNearestTrack(const pat::PackedCandidate* particle, const float minTrackPt=1.0);
+    float computePFIsolation(const pat::PackedCandidate* particle, const float minDR, const float maxDR, const unsigned int isotype=0, const float minNeutralPt=0.5, const float minPUPt=0.5, const float dBeta=0.5);
+    float TrackIso(const pat::PackedCandidate* particle, const float maxDR=0.3, const float deltaZCut=0.1);
+     
+private:
+     
+    virtual void beginStream(edm::StreamID) override;
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+    virtual void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<std::vector<pat::Jet> >             jetToken_;
+    edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pfCandToken_;
+    double        candptMin_;
+    double        candetaMax_;
+};
+
+TauMVAProducer::TauMVAProducer(const edm::ParameterSet& iConfig)
+{
+    //now do what ever other initialization is needed
+    produces<nanoaod::FlatTable>("pfcands");
+    produces<edm::PtrVector<reco::Candidate> >();
+    edm::InputTag jetSrc_    = iConfig.getParameter<edm::InputTag>("jetsSrc");
+    edm::InputTag pfCandSrc_ = iConfig.getParameter<edm::InputTag>("pfCandSrc");
+
+    jetToken_    = consumes<std::vector<pat::Jet> >(jetSrc_);
+    pfCandToken_ = consumes<std::vector<pat::PackedCandidate> >(pfCandSrc_);
+
+    candptMin_ = iConfig.getParameter<double>("minCandPt");
+    candetaMax_ = iConfig.getParameter<double>("maxCandEta");
+}
+
+TauMVAProducer::~TauMVAProducer()
+{
+ 
+}
+
+//
+// member functions
+//
+
+bool TauMVAProducer::isA(int particleID, int pdgId, bool checkCharge)
+{
+  int id = pdgId;
+  if(checkCharge) return (id == particleID);
+
+  return (std::fabs(id) == particleID);
+}
+
+int TauMVAProducer::getContainingJetIndex(const pat::PackedCandidate* pfc) const
+{
+  edm::Handle<std::vector<pat::Jet> > jets_;
```

Comment:
Do you get the jets here? It created an empty handle only. So the loop below won't run no?

Reply:
If you just need to link the PFCandidate to the jet, you can just use example as  https://github.com/susy2015/NanoSUSY/blob/master/plugins/SoftbTableProducer.cc#L163